proposals/NNNN-forbid-weak-keyword-for-computed-properties.md

```diff
@@ -0,0 +1,72 @@
+# Feature name
+
+* Proposal: [SE-NNNN](NNNN-forbid-weak-keyword-for-computed-properties.md)
+* Author: [Andrey Volodin](https://github.com/s1ddok)
+* Status: **Awaiting review**
+* Review manager: TBD
+
+## Introduction
+
+Current compiler version allow `weak` keyword for every type of properties. For stored properties this works just fine, but when it comes to computed ones it brings a lot of questions to the table.
+
+
+## Motivation
+
+Let me start by showing you the most simple example that one can do.
+
+```swift
+class Foo {
+    private var a: Int = 0
+}
+
+class FooFactory {
+    public weak var newFoo: Foo? {
+        return Foo()
+    }
+}
+```
+
+Let's not get into architecture sides of it and proceed straight to the property `newFoo`. As you can see it is a pure computed get-only property. It is very unclear of what `weak` modificator is supposed to mean in this case. As far as I can see it has no affect at all on property's behaviour, but requires to be an optional type. This is strange behaviour which makes code less readable and makes you wonder about how certain things works and what affect some keywords (`weak`) have.
+
+Now we will look at slightly more complex example on how `weak` keyword can bring a lot of confusion to the API.
+
+```swift
+class Foo {
+    private var a: Int = 0
+}
+
+class FooContainer {
+    private var _f: Foo!
+
+    public weak var newFoo: Foo? {
+        get { return _f }
+        set {
+            // some **necessary** logic here, that can't be putted in willSet
+            _f = newValue
+            // some **necessary** logic here, that can't be putted in didSet
+        }
+    }
+}
+```
+
+We all love Swift's `didSet` and `willSet` code containers, they are really useful in avoiding excess code and adding certain behaviour to stored properties. But sometimes we **really** want to implement our own getter/setter. For example if we need to handle both will- and did- set cases, but need to remember some state in between them. As you can see in the example above, our computed (but not a get-only) property is marked by `weak` keyword and has an optional type on it, but the things is that it is mapped to strong private property, which makes the whole thing obsolette.
```

Comment:
We really don't need this much editorializing here (the "we all love..." bits are really distracting for the read). The important point is the last sentence: the property is marked `weak` but actually puts the value into a strong private, stored property, so `weak` is misleading here. That's all you need to say.